tests/L0/run_transformer/test_pipeline_parallel_fwd_bwd.py

```diff
@@ -112,6 +113,17 @@ def _forward_backward_test_impl(
                 deallocate_pipeline_output=deallocate_pipeline_outputs,
             )
 
+            h_size = PipelineParallelForwardBackwardTest.HIDDEN_SIZE
+            m_size = PipelineParallelForwardBackwardTest.MICRO_BATCH_SIZE
+            target_loss = 0
+            for r in range(self.world_size):
+                target_loss = target_loss + h_size*(r + 1) + 1
+            target_loss = target_loss * (h_size ** 2) * m_size / self.world_size
+
+            for loss_item in loss:
+                x = loss_item['avg']
+                self.assertTrue(torch.allclose(x, target_loss*torch.ones_like(x)))
```

Comment:
Currently I think the test runs for multiple iterations, would there need to be some logic to revert the parameters to their previous state due to the call to `optimizer.step()`?

Reply:
The reason of multiple iterations is that we once experienced a hang in the second or later iteration.

For this, an alternative is to store the initial `torch.nn.Module.state_dict` and load it before each forward/backward function call.